src/Swarm/Language/Typecheck.hs

```diff
@@ -508,3 +541,195 @@ check t ty = do
   ty' <- infer t
   _ <- ty =:= ty'
   return ()
+
+-- | Ensure a term is a valid argument to @atomic@.  Valid arguments
+--   may not contain @def@, @let@, or lambda. Any variables which are
+--   referenced must have a primitive, first-order type such as
+--   @string@ or @int@ (in particular, no functions, @cmd@, or
+--   @delay@).  We simply assume that any locally bound variables are
+--   OK without checking their type: the only way to bind a variable
+--   locally is with a binder of the form @x <- c1; c2@, where @c1@ is
+--   some primitive command (since we can't refer to external
+--   variables of type @cmd a@).  If we wanted to do something more
+--   sophisticated with locally bound variables we would have to
+--   inline this analysis into typechecking proper, instead of having
+--   it be a separate, out-of-band check.
+--
+--   The goal is to ensure that any argument to @atomic@ is guaranteed
+--   to evaluate and execute in some small, finite amount of time, so
+--   that it's impossible to write a term which runs atomically for an
+--   indefinite amount of time and freezes the rest of the game.  Of
+--   course, nothing prevents one from writing a large amount of code
+--   inside an @atomic@ block; but we want the execution time to be
+--   linear in the size of the code.
+--
+--   We also ensure that the atomic block takes at most one tick,
+--   i.e. contains at most one tangible command. For example, @atomic
+--   (move; move)@ is invalid, since that would allow robots to move
+--   twice as fast as usual by doing both actions in one tick.
+validAtomic :: Syntax -> Infer ()
+validAtomic s@(Syntax l t) = do
+  n <- analyzeAtomic S.empty s
+  when (n > 1) $ throwError (InvalidAtomic l (TooManyTicks n) t)
+
+-- | Analyze an argument to @atomic@: ensure it contains no nested
+--   atomic blocks and no references to external variables, and count
+--   how many tangible commands it will execute.
+analyzeAtomic :: Set Var -> Syntax -> Infer Int
+analyzeAtomic locals (Syntax l t) = case t of
+  TUnit {} -> return 0
+  -- No nested 'atomic' allowed!
+  TConst Atomic -> throwError $ InvalidAtomic l NestedAtomic t
+  TConst c
+    | c `elem` allowedAtomicConsts -> return $ if isTangible c then 1 else 0
+    | otherwise -> throwError $ InvalidAtomic l (DisallowedConst c) t
+  TDir {} -> return 0
+  TInt {} -> return 0
+  TAntiInt {} -> return 0
+  TString {} -> return 0
+  TAntiString {} -> return 0
+  TBool {} -> return 0
+  TRobot {} -> return 0
+  TRequireDevice {} -> return 0
+  TRequire {} -> return 0
+  SPair s1 s2 -> (+) <$> analyzeAtomic locals s1 <*> analyzeAtomic locals s2
+  -- Special case for if: number of tangible commands is *max* of branches
+  -- instead of sum, since exactly one of them will be executed
+  SApp (STerm (SApp (STerm (SApp (STerm (TConst If)) tst)) thn)) els ->
```

Comment:
`As` should be another exception, it does not really take an extra game time and its arguments should not be checked.

E.g. it took me a minute to go 600 steps with the base until I reached a mountain, but `As` had it in less than a tick.

It is a big responsibility to not freeze the game, but god-like powers should not be limited by mortal concerns. :innocent: 

---

P.S. Do we not have some pattern for binary operators/functions to make this pattern matching shorter?

Reply:
> As should be another exception, it does not really take an extra game time and its arguments should not be checked.

That's... true, I suppose, although it does take *computational* time even if it doesn't take ticks in the game.  And I am not sure why you would need to use `as` inside an `atomic` block.  I am more inclined to leave it as is and revisit this if someone has a compelling use case for a special case later.

Reply:
> Do we not have some pattern for binary operators/functions to make this pattern matching shorter?

We do now! :smile: :heavy_check_mark: